source/fuzz/fact_manager.cpp

```diff
@@ -529,38 +543,41 @@ void FactManager::DataSynonymAndIdEquationFacts::AddFact(
     }
     rhs_dd_ptrs.push_back(synonymous_.Find(&rhs_dd));
   }
-  // We now have the equation in a form where it refers exclusively to
-  // equivalence class representatives.  Add it to our set of facts and work
-  // out any follow-on facts.
-  AddEquationFactRecursive(*lhs_dd_ptr, static_cast<SpvOp>(fact.opcode()),
+  AddEquationFactRecursive(lhs_dd, static_cast<SpvOp>(fact.opcode()),
                            rhs_dd_ptrs, context);
 }
 
 void FactManager::DataSynonymAndIdEquationFacts::AddEquationFactRecursive(
-    const protobufs::DataDescriptor& lhs_dd, SpvOp opcode,
-    const std::vector<const protobufs::DataDescriptor*>& rhs_dds,
+    const protobufs::DataDescriptor& lhs_dd_non_canonical, SpvOp opcode,
+    const std::vector<const protobufs::DataDescriptor*>& rhs_dds_non_canonical,
     opt::IRContext* context) {
-  // Precondition: all data descriptors referenced in this equation must be
-  // equivalence class representatives - i.e. the equation must be in canonical
-  // form.
-  assert(synonymous_.Exists(lhs_dd));
-  assert(synonymous_.Find(&lhs_dd) == &lhs_dd);
-  for (auto rhs_dd : rhs_dds) {
-    (void)(rhs_dd);  // Keep compilers happy in release mode.
-    assert(synonymous_.Exists(*rhs_dd));
-    assert(synonymous_.Find(rhs_dd) == rhs_dd);
+  assert(synonymous_.Exists(lhs_dd_non_canonical) &&
+         "The LHS must be known to the equivalence relation.");
+  // Find the representative for the LHS.
+  auto lhs_dd = synonymous_.Find(&lhs_dd_non_canonical);
+
+  // Find the representative for each RHS operand.
+  std::vector<const protobufs::DataDescriptor*> rhs_dds;
+  for (auto rhs_dd_non_canonical : rhs_dds_non_canonical) {
+    assert(synonymous_.Exists(*rhs_dd_non_canonical) &&
+           "The RHS operands must be known to the equivalence relation.");
+    rhs_dds.push_back(synonymous_.Find(rhs_dd_non_canonical));
   }
 
-  if (id_equations_.count(&lhs_dd) == 0) {
+  // |lhs_dd_non_canonical| and |rhs_dds_non_canonical| should not be used from
+  // here onwards; their canonicalized forms, |lhs_dd| and |rhs_dds|, should be
+  // used instead.
```

Comment:
Is it the case that we could just use `lhs_dd_non_canonical` and `rhs_dds_non_canonical`? But it is likely more efficient to do what you did? And is it true that `lhs_dd` and `rhs_dds` might become "non-canonical" part-way through this function? Could be worth noting this, if true.

Reply:
Actually you are correct that the non-canonical versions can be used, which makes things simpler.  Done.